frontend/ui/widget/textboxwidget.lua

```diff
@@ -11,224 +12,233 @@ A TextWidget that handles long text wrapping
     face = nil,
     bold = nil,
+    line_height = 0.3, -- in em
     fgcolor = Blitbuffer.COLOR_BLACK,
     width = 400, -- in pixels
-    height = nil,
-    first_line = 1,
-    virtual_line = 1, -- used by scroll bar
-    line_height = 0.3, -- in em
-    v_list = nil,
+    height = nil, -- nil value indicates unscrollable text widget
+    virtual_line_num = 1, -- used by scroll bar
     _bb = nil,
-    _length = 0,
 }
 
 function TextBoxWidget:init()
-    local v_list
-    if self.height then
-        v_list = self:_getCurrentVerticalList()
-    else
-        v_list = self:_getVerticalList()
-    end
-    self:_render(v_list)
-    self.dimen = Geom:new(self:getSize())
-end
-
-function TextBoxWidget:_wrapGreedyAlg(h_list)
     local line_height = (1 + self.line_height) * self.face.size
-    local cur_line_width = 0
-    local cur_line = {}
-    local v_list = {}
-
-    for k,w in ipairs(h_list) do
-        w.box = {
-            x = cur_line_width,
-            w = w.width,
+    local font_height = self.face.size
+    self.cursor_line = LineWidget:new{
+        dimen = Geom:new{
+            w = Screen:scaleBySize(1),
             h = line_height,
         }
-        cur_line_width = cur_line_width + w.width
-        if w.word == "\n" then
-            if cur_line_width > 0 then
-                -- hard line break
-                table.insert(v_list, cur_line)
-                cur_line = {}
-                cur_line_width = 0
-            end
-        elseif cur_line_width > self.width then
-            -- wrap to next line
-            table.insert(v_list, cur_line)
-            cur_line = {}
-            cur_line_width = w.width
-            table.insert(cur_line, w)
-        else
-            table.insert(cur_line, w)
-        end
-    end
-    -- handle last line
-    table.insert(v_list, cur_line)
-
-    return v_list
-end
-
-function TextBoxWidget:_getVerticalList(alg)
-    if self.vertical_list then
-        return self.vertical_list
+    }
+    self:_evalCharWidthList()
+    self:_splitCharWidthList()
+    if self.height == nil then
+        self:_renderText(1, #self.vertical_string_list)
+    else
+        self:_renderText(1, self:getVisLineCount())
     end
-    -- build horizontal list
-    local h_list = {}
-    for line in util.gsplit(self.text, "\n", true) do
-        for words in line:gmatch("[\32-\127\192-\255]+[\128-\191]*") do
-            for word in util.gsplit(words, "%s+", true) do
-                for w in util.gsplit(word, "%p+", true) do
-                    local word_box = {}
-                    word_box.word = w
-                    word_box.width = RenderText:sizeUtf8Text(0, Screen:getWidth(), self.face, w, true, self.bold).x
-                    table.insert(h_list, word_box)
-                end
-            end
-        end
-        if line:sub(-1) == "\n" then table.insert(h_list, {word = '\n', width = 0}) end
+    if self.editable then
+        x, y = self:_findCharPos()
```

Comment:
all variables need to be defined as local. We want to avoid defining global variables for better performance and maintainability.